tools/codegen/api/python.py

```diff
@@ -520,12 +702,65 @@ def argument(a: Argument) -> PythonArgument:
         default_init=None,
     )
 
-def signature(f: NativeFunction, *, method: bool = False) -> PythonSignature:
+def argument_type_str_pyi(t: Type) -> str:
+    # vestige from 'dynamic_type': remove optional-ness
+    if isinstance(t, OptionalType):
+        t = t.elem
+
+    if isinstance(t, BaseType):
+        if t.name == BaseTy.int:
+            return '_int'
+        elif t.name == BaseTy.float:
+            return '_float'
+        elif t.name == BaseTy.str:
+            return 'str'
+        elif t.name == BaseTy.Scalar:
+            return 'Number'
+        elif t.name == BaseTy.ScalarType:
+            return '_dtype'
+        elif t.name == BaseTy.bool:
+            return '_bool'
+        elif t.name == BaseTy.QScheme:
+            return '_qscheme'
+        elif t.name == BaseTy.Layout:
+            return '_layout'
+        elif t.name == BaseTy.Device:
+            return 'Union[_device, str, None]'
+        elif t.name == BaseTy.MemoryFormat:
+            return 'memory_format'
+        elif t.name == BaseTy.Dimname:
+            return 'Union[str, ellipsis, None]'
+        elif t.name in [BaseTy.Tensor, BaseTy.Generator,
+                        BaseTy.Storage, BaseTy.Stream, BaseTy.str]:
+            # These python schema type names line up with their function schema names
+            return t.name.name
+
+    elif isinstance(t, ListType):
+        if str(t.elem) == 'int':
+            return 'Union[_int, _size]' if t.size is not None else '_size'
+        elif t.is_tensor_like():
+            return 'Union[Tensor, Tuple[Tensor, ...], List[Tensor]]' if t.size is not None else \
+                   'Union[Tuple[Tensor, ...], List[Tensor]]'
+        elif str(t.elem) == 'float':
+            return 'Sequence[float]'
+        elem = argument_type_str_pyi(t.elem)
+        return f'Sequence[{elem}]'
+
+    raise RuntimeError(f'unrecognized type {repr(t)}')
+
+# Generates a PythonSignature that can be used for either .pyi or PythonArgParser codegen
+def signature(f: NativeFunction, *, method: bool = False, pyi: bool = False) -> PythonSignature:
     # Use cpp api to gather TensorOptions fields from kwargs.
-    # Skip ThisArgument if this is method signature.
+    # Skip SelfArgument if this is method, non-pyi signature.
     # Skip TensorOptionsArguments in C++ signature. Python side TensorOptions
     # arguments are created based on different rules - see below.
-    args = tuple(a for a in cpp.group_arguments(f.func, method=method) if isinstance(a, Argument))
+    cpp_args = cpp.group_arguments(f.func, method=method)
+    args = tuple(a for a in cpp_args if isinstance(a, Argument))
+
+    # pyi method signatures have a SelfArgument, but they reorder it so that it's always the first argument
+    this_args = [a.argument for a in cpp_args if isinstance(a, SelfArgument)]
+    if len(this_args) > 0 and pyi:
+        args = (this_args[0],) + args
```

Comment:
I'd probably suggest doing this slightly differently. Instead of relying on args being populated with the self argument in the signature, at the site where you are consuming PythonSignature to generate methods, just fill in self at that point. The merit of this approach is that you no longer need to vary the behavior of this function based on pyi; just unconditionally strip self argument, and then add it back in when you know you need it.

Reply:
I like this idea as well.
@bdhirsh similarly, I wonder if it is possible to have different sets of 'presenter' methods (argument_str, argument_str_pyi, signature_str v.s. signature_str_pyi, etc) without persisting the pyi bool in the data model - the  main codegen script should know which sets of presenters to use, right?

Reply:
Yeah, I think that nets out to being cleaner.

The downside of that is that the state of the dataclass isn't really correct anymore- it feels a little gross to from NativeFunction (including `self`) -> PythonSignature (excluding `self`) -> printed signature string (including `self`), where `self` gets tacked onto the signature string at the very end. I was trying to prefer keeping the state of  the PythonSignature to be as representative as possible.

But yeah, we're also kind of bending over backwards to deal with weirdness (self only shows up for pyi, and it gets reordered), so I'm fine with doing that

Reply:
@ljk53 yeah, that should be pretty easy to switch to. I'll try adding that

Reply:
@bdhirsh thanks! additional nits: if you can group the pyi presenters together under one section it would be clear that it's different from the python->c++ binding logic that the rest of the source file does...